File: 889-construct-binary-tree-from-preorder-and-postorder-traversal/889-construct-binary-tree-from-preorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] preorder;
    int[] postorder;
    Map<Integer,Integer> postMap;
    public TreeNode constructFromPrePost(int[] preorder, int[] postorder) {
        this.preorder = preorder;
        this.postorder = postorder;
        int n = preorder.length;
        postMap = new HashMap<>();
        for(int i = 0; i < n; i++) postMap.put(postorder[i], i);
        return dfs(0, n - 1, 0, n - 1);
    }
    public TreeNode dfs(int prestart, int preend, int poststart, int postend) {
        if(prestart > preend || poststart > postend) return null;
        if(prestart == preend) return new TreeNode(preorder[prestart]);
        int rootVal = preorder[prestart];
        int leftVal = preorder[prestart + 1];
        int leftIdx = postMap.get(leftVal);
        int leftLen = leftIdx - poststart;
        TreeNode root = new TreeNode(rootVal);
        root.left = dfs(prestart + 1, prestart + leftLen + 1, poststart, leftIdx);
        root.right = dfs(prestart + leftLen + 2, preend, leftIdx + 1, postend - 1);
        return root;
    }
}

File: 889-construct-binary-tree-from-preorder-and-postorder-traversal/889-construct-binary-tree-from-preorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        n = len(preorder)
        preorder_map = {}
        for i, val in enumerate(preorder):
            preorder_map[val] = i
        def buildTree(prestart, preend, poststart, postend):
            nonlocal preorder_map
            if prestart > preend or poststart > postend:
                return None
            print(prestart)
            root = TreeNode(preorder[prestart])
            if(prestart == preend): return root;
            
            left_count = preorder_map[postorder[postend - 1]] - prestart - 1
            if(left_count < 0):
                left_count = 0
            root.left = buildTree(prestart + 1, prestart + left_count, poststart, poststart + left_count - 1)
            root.right = buildTree(prestart + left_count + 1, preend, poststart + left_count, postend - 1)
            return root
        return buildTree(0, n - 1, 0, n - 1)
    
            